### PR TITLE
[WIP] Define IRuntimeVariables in references.

### LINKED ISSUE #8067: ProjectReference -> Package Dependency: Disabled System.Linq.Expressions from package dependency build
System.Linq.Expressions does not build when project references are converted to package dependencies due to 

> Catalog\ExpressionCatalog.RuntimeVariables.cs(34,39): error CS0246: The type or namespace name 'IRuntimeVariables' could not be found (are you missing a using directive or an assembly reference?) [D:\A_work\8\s\src\System.Linq.Expressions\tests

See PR #7998

### PR DESCRIPTION
**Pull request by JonHanna**
Fixes #8067

### CODE REVIEW

**Comment by JonHanna** [comment]
At least I think it might. Will discuss at that issue.

**Comment by JonHanna** [comment]
Test  Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test
(Permissions)

**Comment by JonHanna** [comment]
Test Innerloop OSX Debug Build and Test

File: src/System.Linq.Expressions/ref/System.Linq.Expressions.cs [hunk 1]
```diff
@@ -841,3 +841,24 @@ public sealed partial class UnaryExpression : System.Linq.Expressions.Expression
...
+        int Count { get; }
+
+        /// <summary>
+        /// An indexer to get/set the values of the runtime variables.
+        /// </summary>
+        /// <param name="index">An index of the runtime variable.</param>
+        /// <returns>The value of the runtime variable.</returns>
+        object this[int index] { get; set; }
+    }
+}
```
Comment (jasonwilliams200OK) [inline src/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: tiny nit: missing the end of file linebreak.

Comment (JonHanna) [inline src/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Fixed. Thanks.

Comment (JonHanna) [inline src/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Hmm. Curious and annoying. The commit I have here definitely includes the line break at the end. My local copy definitely shows it, and yet the view through github here does not.

**Comment by chcosta** [comment]
I cloned your branch to validate your change.  Here is the process I followed (after cloning your branch):

1. build.cmd /p:BuildTestsAgainstPackages=true
    a. This step builds the entire repo, but for test projects, it removes the project's ProjectReference items and generates a new project.json with the project references added as dependencies (these project.json's are generated into bin\obj\generated).
    b. This step failed with the same error we were previously seeing but this is because the publically available packages don't have your fix.  I let the build complete, though, so it would build packages.  To validate your fix, I then...
2. pushd src\System.Linq.Expressions\tests
3. msbuild /p:BuildTestsAgainstPackages=true /p:PackagesDrops=[bindir]\packages /p:PackagesBuildNumberOverride=rc3-24109-0
  a. This tells the build to use packages from "bin\packages", and sets the version number to what I just built so we are using the packages built with your change.

The result, is a few more errors.  I'm not certain if the correct thing to do is to continue to iterate and alter the public / private surface area, or something else.  @weshaggard may have additional insight.

```
  ExpressionTests.cs(119,19): error CS1729: 'Expression' does not contain a constructor that takes 2 arguments [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(170,34): error CS1061: 'ElementInit' does not contain a definition for 'ArgumentCount' and no extension method 'ArgumentCount' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(171,35): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(172,37): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(182,76): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(183,76): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(184,76): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ElementInitTests.cs(185,76): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
  ListInit\ListInitExpressionTests.cs(134,75): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found (are you missing a using directive or an assembly reference?) [e:\gh\jonhanna\corefx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
```

**Comment by JonHanna** [comment]
> ExpressionTests.cs(119,19): error CS1729: 'Expression' does not contain a constructor that takes 2 arguments

Hmm. That it most certainly does, but it's obsolete, and its obsolete for a good reason. It would be nice if people would all stop using it, so I wonder if this was deliberately left out of the ref interface. It should though probably be included.

> ListInit\ElementInitTests.cs(170,34): error CS1061: 'ElementInit' does not contain a definition for 'ArgumentCount' and no extension method 'ArgumentCount' accepting a first argument of type 'ElementInit' could be found

>  ListInit\ElementInitTests.cs(171,35): error CS1061: 'ElementInit' does not contain a definition for 'GetArgument' and no extension method 'GetArgument' accepting a first argument of type 'ElementInit' could be found

These are both implementations of `IArgumentProvider` that somehow became implicit rather than explicit between the last framework version (and the reference source) and the first import into CoreFX. I've come across a few of these before which were more blatantly bugs and so I changed them back rather than wrote tests for them like I did here!

They should probably be returned to being explicit, and the test changed to testing as such, but it's not a clear case.

**Comment by stephentoub** [comment]
@JonHanna, what's the status of this?  Can it be closed as a PR while the issue is discussed?

**Comment by JonHanna** [comment]
It may as well if only because I'm not able to do much right now.

**Pull request closed without merging**